packages/zoe/src/zoeService/zoeSeat.js

```diff
@@ -176,33 +191,59 @@ export const makeZoeSeatAdminFactory = baggage => {
           );
         },
         // called only for seats resulting from offers.
-        resolveExitAndResult(offerResultPromise, exitObj) {
+        /** @param {HandleOfferResult} result */
+        resolveExitAndResult({ offerResultPromise, exitObj }) {
           const { state, facets } = this;
 
-          if (ephemeralOfferResultStore.has(facets.userSeat)) {
-            const pKit = ephemeralOfferResultStore.get(facets.userSeat);
-            E.when(
-              offerResultPromise,
-              offerResult => {
-                pKit.resolve(offerResult);
-                state.offerResult = offerResult;
-                state.offerResultValid = true;
-                ephemeralOfferResultStore.delete(facets.userSeat);
-              },
-              e => {
-                pKit.reject(e);
-                state.offerResult = pKit.promise;
-                state.offerResultValid = true;
-                ephemeralOfferResultStore.delete(facets.userSeat);
-              },
-            );
-          } else {
+          !state.offerResultStored ||
+            Fail`offerResultStored before offerResultPromise`;
+
+          if (!ephemeralOfferResultStore.has(facets.userSeat)) {
+            // this was called before getOfferResult
             const kit = makePromiseKit();
             kit.resolve(offerResultPromise);
             ephemeralOfferResultStore.set(facets.userSeat, kit);
-            state.offerResultValid = false;
           }
 
+          const pKit = ephemeralOfferResultStore.get(facets.userSeat);
+          E.when(
+            offerResultPromise,
+            offerResult => {
+              // Resolve the ephemeral promise for offerResult
+              pKit.resolve(offerResult);
+              // Now we want to store the offerResult in `state` to get it off the heap,
+              // but we need to handle three cases:
+              //   1. already durable. (This includes being a remote presence.)
+              //   2. has promises for durable objects.
+              //   3. not durable even after resolving promises.
+              // For #1 we can assign directly, but we deeply await to also handle #2.
+              void E.when(
+                deeplyFulfilled(offerResult),
```

Comment:
Is this safe if the object is not a record?

I had my self gone for a deeply fulfilled, but more upstream in vaults only. See https://github.com/Agoric/agoric-sdk/commit/666f0825b22b15c473b760971028923a12371e1c

Reply:
> Is this safe if the object is not a record?

looks like it; `deeplyFulfilled` starts with:

```
  if (!isObject(val)) {
    return val;
  }
```

Reply:
Sorry I meant copyRecord / copyArray, aka a presence or remotable. I guess yes since it already leaves these in place?

Reply:
OIC... but looking closer, yes, `deeplyFulfilled` clearly preserves the passstyle at each level. For example, [`copyArray`](https://github.com/endojs/endo/blob/916f6d42270b282641d3299abdb4ca9c835a039b/packages/marshal/src/deeplyFulfilled.js#L60).